libkbfs/file_data.go

```diff
@@ -750,54 +1060,120 @@ func (fd *fileData) truncateExtend(ctx context.Context, size uint64,
-	// We need to delete some data (and possibly entire blocks).
+	// We need to delete some data (and possibly entire blocks).  Note
+	// we make a new slice and copy data in order to make sure the
+	// data being truncated can be fully garbage-collected.
 	block.Contents = append([]byte(nil), block.Contents[:iSize-startOff]...)
 
 	newlyDirtiedChildBytes = int64(len(block.Contents))
 	if wasDirty {
 		newlyDirtiedChildBytes -= int64(oldLen) // negative
 	}
 
+	dirtyMap := make(map[BlockPointer]bool)
 	if nextBlockOff > 0 {
-		// TODO: if indexInParent == 0, we can remove the level of indirection.
-		// TODO: support multiple levels of indirection.
-		parentBlock := parentBlocks[0].pblock
-		indexInParent := parentBlocks[0].childIndex
-		for _, ptr := range parentBlock.IPtrs[indexInParent+1:] {
-			unrefs = append(unrefs, ptr.BlockInfo)
+		// TODO: remove any unnecessary levels of indirection if the
+		// number of leaf nodes shrinks significantly (KBFS-1824).
+
+		// Get all paths to any leaf nodes following the new
+		// right-most block, since those blocks need to be
+		// unreferenced, and their parents need to be modified or
+		// unreferenced.
+		pfr, err := fd.getIndirectBlocksForOffsetRange(
+			ctx, topBlock, nextBlockOff, -1)
+		if err != nil {
+			return DirEntry{}, nil, nil, 0, err
 		}
-		parentBlock.IPtrs = parentBlock.IPtrs[:indexInParent+1]
-		// always make the parent block dirty, so we will sync it
-		if err = fd.cacher(fd.file.tailPointer(), parentBlock); err != nil {
-			return DirEntry{}, nil, newlyDirtiedChildBytes, err
+
+		cachedPtrs := make(map[BlockPointer][]IndirectFilePtr)
+		for _, path := range pfr {
+			// Fake an indirect pointer for the top block.
+			iptr := IndirectFilePtr{BlockInfo: BlockInfo{
+				BlockPointer: fd.rootBlockPointer(),
+			}}
+			leftMost := true
+			for i, pb := range path {
+				ptrs := cachedPtrs[iptr.BlockPointer]
+				if ptrs == nil {
+					// Process each block exactly once, removing all
+					// now-unnecessary indirect pointers (but caching
+					// that list so we can still walk the tree on the
+					// next iterations).
+					pblock := pb.pblock
+					ptrs = pblock.IPtrs
+					cachedPtrs[iptr.BlockPointer] = ptrs
+
+					// Remove the first child iptr and everything
+					// following it if all the child indices below
+					// this level are 0. If we remove iptr 0, this
+					// block can be unreferenced (unless it's on the
+					// left-most edge of the tree, in which case we
+					// keep it around for now -- see above TODO).
+					removeStartingFromIndex := pb.childIndex
+					for j := i + 1; j < len(path); j++ {
+						if path[j].childIndex > 0 {
+							removeStartingFromIndex++
+							break
+						}
+					}
+
+					if pb.childIndex == 0 && !leftMost {
+						if iptr.EncodedSize != 0 {
+							unrefs = append(unrefs, iptr.BlockInfo)
+						}
+					} else if removeStartingFromIndex < len(pblock.IPtrs) {
+						pblock.IPtrs = pblock.IPtrs[:removeStartingFromIndex]
```

Comment:
I'm worried about this, because it seems to me this will modify a read-only block -- pblock is a pointer, and it's retrieved with `getLeafBlocksForOffsetRange`, which uses `blockReadParallel`, which doesn't make a copy.